test/jdk/java/foreign/TestAddressDereference.java

```diff
@@ -0,0 +1,156 @@
+                    MethodType.methodType(void.class, MemorySegment.class, long.class));
+        } catch (Throwable ex) {
+            throw new AssertionError(ex);
+        }
+    }
+
+    @Test(dataProvider = "layoutsAndAlignments")
+    public void testGetAddress(long alignment, ValueLayout layout) {
+        boolean badAlign = layout.byteAlignment() > alignment;
+        try (Arena arena = Arena.openConfined()) {
+            MemorySegment segment = arena.allocate(ValueLayout.ADDRESS);
+            segment.set(ValueLayout.ADDRESS, 0, MemorySegment.ofAddress(alignment));
+            MemorySegment deref = segment.get(ValueLayout.ADDRESS.withTargetLayout(layout), 0);
+            assertFalse(badAlign);
+            assertEquals(deref.byteSize(), layout.byteSize());
+        } catch (IllegalArgumentException ex) {
+            assertTrue(badAlign);
+            assertTrue(ex.getMessage().contains("alignment constraint for address"));
+        }
+    }
+
+    @Test(dataProvider = "layoutsAndAlignments")
+    public void testGetAddressIndex(long alignment, ValueLayout layout) {
+        boolean badAlign = layout.byteAlignment() > alignment;
+        try (Arena arena = Arena.openConfined()) {
+            MemorySegment segment = arena.allocate(ValueLayout.ADDRESS);
+            segment.set(ValueLayout.ADDRESS, 0, MemorySegment.ofAddress(alignment));
+            MemorySegment deref = segment.getAtIndex(ValueLayout.ADDRESS.withTargetLayout(layout), 0);
+            assertFalse(badAlign);
+            assertEquals(deref.byteSize(), layout.byteSize());
+        } catch (IllegalArgumentException ex) {
+            assertTrue(badAlign);
+            assertTrue(ex.getMessage().contains("alignment constraint for address"));
+        }
+    }
+
+    @Test(dataProvider = "layoutsAndAlignments")
+    public void testNativeReturn(long alignment, ValueLayout layout) throws Throwable {
+        boolean badAlign = layout.byteAlignment() > alignment;
+        try {
+            MethodHandle get_addr_handle = LINKER.downcallHandle(GET_ADDR_SYM,
+                    FunctionDescriptor.of(ValueLayout.ADDRESS.withTargetLayout(layout), ValueLayout.ADDRESS));
+            MemorySegment deref = (MemorySegment)get_addr_handle.invokeExact(MemorySegment.ofAddress(alignment));
+            assertFalse(badAlign);
+            assertEquals(deref.byteSize(), layout.byteSize());
+        } catch (IllegalArgumentException ex) {
+            assertTrue(badAlign);
+            assertTrue(ex.getMessage().contains("alignment constraint for address"));
+        }
+    }
+
+    @Test(dataProvider = "layoutsAndAlignments")
+    public void testNativeUpcallArg(long alignment, ValueLayout layout) throws Throwable {
+        boolean badAlign = layout.byteAlignment() > alignment;
+        if (badAlign) return; // this will crash the JVM (exception occurs when going into the upcall stub)
```

Comment:
Any idea on this @JornVernee ? When the alignment exception is thrown (this happens when we execute the box address bindings before entering the upcall stub), the JVM exits, reporting an exception that points to where the downcall is made, which is quite confusing. I still think the check is worthwhile having (after all a layout is more than just the size), it's a bit unfortunate that in the case of upcalls a failed check ends up terminating the JVM (but then again, if you are really expecting a pointer with certain characteristics, and you are given another - you might just cross fingers and hope it would be ok).